document JB IDE settings configuration (#3030)
gitpod-io/website: generators/ide/jb-product.md

It is not yet possible to install plugins on **{title} backend** for your user to share across all your Gitpod workspaces.

## Configure JVM options

> Configuration of JVM options is currently in [Alpha](/docs/help/public-roadmap/release-cycle) · [Send feedback](https://github.com/gitpod-io/gitpod/issues/8704).

You can adjust JVM options for {title} backend, especially if you want to increase the `-Xmx` memory size. For example:

```yaml
jetbrains:
  __productId__:
    vmoptions: "-Xmx4g"
```

For more detailed information on JVM options, refer to [Common JVM Options](https://www.jetbrains.com/help/__productDocCode__/tuning-the-ide.html#common-jvm-options) from JetBrains documentation.

## Configure IDE settings

> > As a medium-term workaround, Gitpod allows preconfiguring global settings in predefined locations. If you are using this workaround please leave your use case on [6576](https://github.com/gitpod-io/gitpod/issues/6576). JetBrains is looking for [the long-term solution](https://youtrack.jetbrains.com/issue/RDCT-1/Settings-synchronization) for remote development. Feel free to leave your feedback there as well.

Reviewer: * **Thought/suggestion:** The phrase "Medium term workaround" feels vague, what message do we mean to share? Looks like a request for feedback, maybe adding "feedback required" or similar to the start would help be more clear about the call to action?
* **Thought:** Is it "preconfigure" or simply "configure"?
* **Thought** Would mentioning the JB partnership help here? It would set the context for our ask for feedback?

```suggestion
> **Feedback needed:** JetBrains Gateway functionality and how JetBrains IDE settings work is evolving. [Gitpod continues to collaborate with JetBrains](https://www.gitpod.io/blog/gitpod-jetbrains) on the future of remote development with JetBrains IDEs. 

Gitpod allows you to configure your global settings in predefined locations. If you are using this workaround please leave your use case on [6576](https://github.com/gitpod-io/gitpod/issues/6576). JetBrains is looking for [the long-term solution](https://youtrack.jetbrains.com/issue/RDCT-1/Settings-synchronization) for remote development. Feel free to leave your feedback there as well.
```

> {title} applies [global settings](https://www.jetbrains.com/help/__productDocCode__/configuring-project-and-ide-settings.html) to all projects on your machine. In a remote scenario, some settings are configured on **JetBrains Client** and others on **{title} backend**.

Reviewer: * **Thought:** The phrasing "toi all projects on your machine" seems confusing to me in this context. With Gitpod we are telling users that they're not coding on their machine, is there any way we could phrase this differently or more accurately? 

```suggestion
{title} applies [global settings](https://www.jetbrains.com/help/__productDocCode__/configuring-project-and-ide-settings.html) to all projects on your machine. In a remote scenario, when using JetBrains Gateway, some IDE settings are configured on the **JetBrains Client** and others on the **{title} backend**.
```

### On JetBrains Client

Settings configured on JetBrains Client are stored on your local machine and don't have `On Host` label in the IDE settings.

> They are reused if the workspace is running the exact same version of the {title} backend where settings were initially configured.

Reviewer: > They are reused if the workspace is running the exact same version of the {title} backend where settings were initially configured.

**Question:** This implies that these client settings are wiped for every backend version update, is that correct? Given that we (Gitpod) manage the versions of the backend version, should we also mention that here? e.g. the Gitpod release cycle and how often those settings are wiped / will have to be restored?

<figure>

> <img class="shadow-medium w-full rounded-xl max-w-3xl mt-x-small" alt="Configure IDE settings on JetBrains client" src="/images/editors/jb-client-configure-settings.webp">

Reviewer: **Thought:** This image doesn't show an "onHost" setting, would it make sense to include a screenshot that does? Or even both? One with an `onHost` label and one without? Or have a side-by-side of the images? I found myself looking at the image and searching to see the "onHost" label in the UI, but couldn't find it (naturally).

Reviewer 2: Image updated

>     <figcaption>JetBrains client settings don't have `On Host` label.</figcaption>

Reviewer: Needs to be HTML, not markdown for `code` formatting 🙏 

```suggestion
    <figcaption>JetBrains client settings don't have <code>On Host</code> label.</figcaption>
```

</figure>

### On {title} backend

Settings configured on {title} backend are stored in a Gitpod workspace and have `On Host` label in the IDE settings.

<figure>
<img class="shadow-medium w-full rounded-xl max-w-3xl mt-x-small" alt="Configure IDE settings on JetBrains client" src="/images/editors/jb-backend-configure-settings.webp">
    <figcaption>{title} backend settings have `On Host` label.</figcaption>
</figure>

> #### Preconfigure for the project

Reviewer: **Suggestion:** This title seems a little ambiguous to me, could we rename? e.g. we say "pre" in "preconfigure", but what does "pre" stand for here? I assume "pre" workspace start? If so, could we rename to: 

```suggestion
#### Configure IDE settings for new workspaces
```

> You may want to preconfigure such settings to apply for each newly started Gitpod workspace. Additionally, depending on the use case you may want to configure different settings for different scopes, e.g. for all your workspaces, project, team, or the entire organization.

Reviewer: ```suggestion
You may want to preconfigure IDE settings to apply for each newly started Gitpod workspace. Depending on your use case you may want to additionally configure settings for different scopes, e.g. apply settings to all workspaces, project, team, or the entire organization.
```

> Unfortunately, some settings cannot be configured via [environment variables](/docs/configure/projects/environment-variables#environment-variables) or [vm options](#configure-jvm-options), only via the UI of JetBrains client.

Reviewer: ```suggestion
Some settings can only be set via the UI of JetBrains client, and cannot be configured either via [environment variables](/docs/configure/projects/environment-variables#environment-variables) or [vm options](#configure-jvm-options), meaning there is no way to persist some settings between Gitpod workspace starts. 
```

| Predefined location                                                                                                          | JetBrains global settings for                                    |
| ---------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------- |
| `/home/gitpod/.gitpod/jetbrains/options`                                                                                     | Each JetBrains product configured in a Docker image.             |
| `/home/gitpod/.gitpod/jetbrains/__productId__/options`                                                                       | {title} backend configured in a Docker image.                    |

> | `/{project}/.gitpod/jetbrains/options`<br>Example: `/workspace/gitpod/.gitpod/jetbrains/options`                             | Each JetBrains product configured in a Git repo of your project. |
> | `/{project}/.gitpod/jetbrains/__productId__/options`<br>Example: `/workspace/gitpod/.gitpod/jetbrains/__productId__/options` | {title} backend configured in a Git repo of the project.         |

Reviewer 3: Maybe we can use something other than `gitpod` in the example to avoid the repeated name and to make it more transparent where the project name enters.

```suggestion
| `/{project}/.gitpod/jetbrains/options`<br>Example: `/workspace/my-awesome-project/.gitpod/jetbrains/options`                             | Each JetBrains product configured in a Git repo of your project. |
| `/{project}/.gitpod/jetbrains/__productId__/options`<br>Example: `/workspace/my-awesome-project/.gitpod/jetbrains/__productId__/options` | {title} backend configured in a Git repo of the project.         |
```

> | `/{project}/.gitpod/jetbrains/__productId__/options`<br>Example: `/workspace/gitpod/.gitpod/jetbrains/__productId__/options` | {title} backend configured in a Git repo of the project.         |

Reviewer: **Thought:** I'm not sure I fully understand this table... will need to ask some questions...
